// elm-watch-lib@1.0.1/index.js
// src/NonEmptyArray.ts
import * as Codec from "tiny-decoders";
function NonEmptyArray(decoder) {
  return Codec.flatMap(Codec.array(decoder), {
    decoder: (array3) => isNonEmptyArray(array3) ? { tag: "Valid", value: array3 } : {
      tag: "DecoderError",
      error: {
        tag: "custom",
        message: "Expected a non-empty array",
        got: array3,
        path: []
      }
    },
    encoder: (array3) => array3
  });
}
function isNonEmptyArray(array3) {
  return array3.length >= 1;
}
function mapNonEmptyArray(array3, f) {
  return array3.map(f);
}

// src/SpawnElm.ts
import * as path2 from "path";
import * as Codec9 from "tiny-decoders";

// src/ElmMakeError.ts
import * as Codec3 from "tiny-decoders";

// src/Types.ts
import * as Codec2 from "tiny-decoders";
function brand() {
  return Codec2.string;
}
var AbsolutePath = brand();
function markAsAbsolutePath(string7) {
  return string7;
}
var CompilationMode = Codec2.primitiveUnion([
  "debug",
  "standard",
  "optimize"
]);
var BrowserUiPosition = Codec2.primitiveUnion([
  "TopLeft",
  "TopRight",
  "BottomLeft",
  "BottomRight"
]);
function markAsElmJsonPath(absolutePath) {
  return absolutePath;
}
function markAsSourceDirectory(absolutePath) {
  return absolutePath;
}
var TargetName = brand();
var WebSocketToken = brand();

// src/ElmMakeError.ts
var Color = Codec3.primitiveUnion([
  "red",
  "RED",
  "magenta",
  "MAGENTA",
  "yellow",
  "YELLOW",
  "green",
  "GREEN",
  "cyan",
  "CYAN",
  "blue",
  "BLUE",
  "black",
  "BLACK",
  "white",
  "WHITE"
]);
var StyledText = Codec3.fields({
  bold: Codec3.boolean,
  underline: Codec3.boolean,
  color: Codec3.nullOr(Color),
  string: Codec3.string
});
var MessageChunk = Codec3.flatMap(Codec3.multi(["string", "object"]), {
  decoder: (value) => {
    switch (value.type) {
      case "string":
        return {
          tag: "Valid",
          value: value.value
        };
      case "object":
        return StyledText.decoder(value.value);
    }
  },
  encoder: (value) => typeof value === "string" ? {
    type: "string",
    value
  } : {
    type: "object",
    value: StyledText.encoder(value)
  }
});
var Position = Codec3.fields({
  line: Codec3.number,
  column: Codec3.number
});
var Region = Codec3.fields({
  start: Position,
  end: Position
});
var Problem = Codec3.fields({
  title: Codec3.string,
  region: Region,
  message: Codec3.array(MessageChunk)
});
var CompileError = Codec3.fields({
  // https://github.com/elm/compiler/blob/94715a520f499591ac6901c8c822bc87cd1af24f/compiler/src/Reporting/Error.hs#L42
  path: AbsolutePath,
  name: Codec3.string,
  problems: NonEmptyArray(Problem)
});
var GeneralErrorPath = Codec3.nullOr(Codec3.primitiveUnion(["elm.json"]));
var ElmMakeError = Codec3.taggedUnion("type", [
  {
    type: Codec3.tag("error"),
    // `Nothing` and `Just "elm.json"` are the only values I’ve found in the compiler code base.
    path: GeneralErrorPath,
    title: Codec3.string,
    message: Codec3.array(MessageChunk)
  },
  {
    type: Codec3.tag("compile-errors"),
    errors: NonEmptyArray(CompileError)
  }
]);

// src/Env.ts
var __ELM_WATCH_ELM_TIMEOUT_MS = "__ELM_WATCH_ELM_TIMEOUT_MS";

// src/Errors.ts
import * as crypto from "crypto";
import * as fs2 from "fs";
import * as Codec8 from "tiny-decoders";
import * as url from "url";

// src/ElmWatchJson.ts
import * as Codec7 from "tiny-decoders";

// src/IsWindows.ts
import * as os from "os";
var IS_WINDOWS = os.platform() === "win32";

// src/PathHelpers.ts
import * as fs from "fs";
import * as path from "path";
import * as Codec5 from "tiny-decoders";

// src/Helpers.ts
import * as Codec4 from "tiny-decoders";
function split(string7, splitter) {
  return string7.split(splitter);
}
var RESET_COLOR = "\x1B[0m";
function bold(string7) {
  return `${RESET_COLOR}\x1B[1m${string7}${RESET_COLOR}`;
}
function dim(string7) {
  return `${RESET_COLOR}\x1B[2m${string7}${RESET_COLOR}`;
}
function silentlyReadIntEnvValue(value, defaultValue) {
  return /^\d+$/.test(value ?? "") ? Number(value) : defaultValue;
}
var toError = (arg) => (
  /* v8 ignore start */
  arg instanceof Error ? arg : new Error(
    `Caught error not instanceof Error: ${unknownErrorToString(arg)}`
  )
);
function unknownErrorToString(error) {
  return typeof error?.stack === "string" ? error.stack : typeof error?.message === "string" ? error.message : Codec4.repr(error);
}

// src/PathHelpers.ts
function absolutePathFromString(from, ...pathStrings) {
  return markAsAbsolutePath(path.resolve(from, ...pathStrings));
}
function absoluteDirname(absolutePath) {
  return markAsAbsolutePath(path.dirname(absolutePath));
}
function readJsonFile(file, codec) {
  let content;
  try {
    content = fs.readFileSync(file, "utf8");
  } catch (error) {
    return { tag: "ReadError", error: toError(error) };
  }
  return Codec5.JSON.parse(codec, content);
}

// src/Port.ts
import * as Codec6 from "tiny-decoders";
function markAsPort(number3) {
  return number3;
}
var Port = Codec6.flatMap(Codec6.number, {
  decoder: (number3) => {
    const min = 1;
    const max = 65535;
    return Number.isInteger(number3) && min <= number3 && number3 <= max ? {
      tag: "Valid",
      value: markAsPort(number3)
    } : {
      tag: "DecoderError",
      error: {
        tag: "custom",
        message: `Expected an integer where ${min} <= port <= ${max}`,
        got: number3,
        path: []
      }
    };
  },
  encoder: (port) => port
});

// src/ElmWatchJson.ts
var INPUT_NAME = /(^|[/\\])\p{Lu}[_\d\p{L}]*\.elm$/u;
function isValidInputName(name) {
  return INPUT_NAME.test(name);
}
function isValidOutputName(name) {
  return !name.startsWith("-") && name.endsWith(".js") && name !== ".js";
}
var TARGET_NAME = /^[^\s-](?:.*\S)?$/;
function isValidTargetName(name) {
  return TARGET_NAME.test(name);
}
var Target = Codec7.fields(
  {
    inputs: NonEmptyArray(
      Codec7.flatMap(Codec7.string, {
        decoder: (string7) => isValidInputName(string7) ? { tag: "Valid", value: string7 } : {
          tag: "DecoderError",
          error: {
            tag: "custom",
            message: "Inputs must have a valid module name and end with .elm",
            got: string7,
            path: []
          }
        },
        encoder: (value) => value
      })
    ),
    output: Codec7.flatMap(Codec7.string, {
      decoder: (output) => isValidOutputName(output) ? { tag: "Valid", value: output } : {
        tag: "DecoderError",
        error: {
          tag: "custom",
          message: "Outputs must end with .js",
          got: output,
          path: []
        }
      },
      encoder: (value) => value
    })
  },
  { allowExtraFields: false }
);
var TargetRecordHelper = {
  decoder: (record2) => {
    const keys = Object.keys(record2);
    for (const key of keys) {
      if (!isValidTargetName(key)) {
        return {
          tag: "DecoderError",
          error: {
            tag: "custom",
            message: "Target names must start with a non-whitespace character except `-`,\ncannot contain newlines and must end with a non-whitespace character",
            got: key,
            path: [key]
          }
        };
      }
    }
    return isNonEmptyArray(keys) ? { tag: "Valid", value: record2 } : {
      tag: "DecoderError",
      error: {
        tag: "custom",
        message: "Expected a non-empty object",
        got: record2,
        path: []
      }
    };
  },
  encoder: (value) => value
};
var Config = Codec7.fields(
  {
    targets: Codec7.flatMap(Codec7.record(Target), TargetRecordHelper),
    postprocess: Codec7.field(NonEmptyArray(Codec7.string), { optional: true }),
    port: Codec7.field(Port, { optional: true })
  },
  { allowExtraFields: false }
);

// src/Logger.ts
var DEFAULT_COLUMNS = 80;

// src/Theme.ts
var THEME_ESCAPES_STRING = Array.from(
  { length: 16 },
  (_, i) => `\x1B]4;${i};?\x1B\\`
  // Palette.
).concat("\x1B]10;?\x1B\\", "\x1B]11;?\x1B\\").join("");
var COLOR_TO_TERMINAL_ESCAPE = {
  red: "\x1B[31m",
  RED: "\x1B[91m",
  magenta: "\x1B[35m",
  MAGENTA: "\x1B[95m",
  yellow: "\x1B[33m",
  YELLOW: "\x1B[93m",
  green: "\x1B[32m",
  GREEN: "\x1B[92m",
  cyan: "\x1B[36m",
  CYAN: "\x1B[96m",
  blue: "\x1B[34m",
  BLUE: "\x1B[94m",
  black: "\x1B[30m",
  BLACK: "\x1B[90m",
  white: "\x1B[37m",
  WHITE: "\x1B[97m"
};

// src/Errors.ts
function bold2(string7) {
  return { tag: "Bold", text: string7 };
}
function dim2(string7) {
  return { tag: "Dim", text: string7 };
}
function text(string7) {
  return { tag: "Text", text: string7.trim() };
}
var elmJson = bold2("elm.json");
var elmWatchJson = bold2("elm-watch.json");
var elmWatchStuffJson = bold2("elm-stuff/elm-watch/stuff.json");
var fancyError = (title, location) => (strings, ...values) => (width, renderPiece) => ({
  title,
  location: fancyToPlainErrorLocation(location),
  content: template(strings, ...values)(width, renderPiece)
});
var template = (strings, ...values) => (width, renderPiece) => strings.flatMap((string7, index) => {
  const value = values[index] ?? text("");
  return [
    string7,
    typeof value === "function" ? value(width, renderPiece) : renderPiece(value)
  ];
}).join("").trim();
function toTerminalString(errorTemplate, width, noColor) {
  const renderPiece = noColor ? (piece) => piece.text : renderPieceForTerminal;
  const { title, location, content } = errorTemplate(width, renderPiece);
  const prefix = `-- ${title} `;
  const line = "-".repeat(Math.max(0, width - prefix.length));
  const titleWithSeparator = renderPiece(bold2(`${prefix}${line}`));
  return [
    titleWithSeparator,
    ...location === void 0 ? [] : [renderPiece(renderErrorLocation(location))],
    "",
    content
  ].join("\n");
}
function toPlainString(errorTemplate) {
  return toTerminalString(errorTemplate, DEFAULT_COLUMNS, true);
}
function renderPieceForTerminal(piece) {
  switch (piece.tag) {
    case "Bold":
      return bold(piece.text);
    case "Dim":
      return dim(piece.text);
    case "ElmStyle":
      return (piece.bold ? (
        /* v8 ignore next */
        "\x1B[1m"
      ) : "") + (piece.underline ? "\x1B[4m" : "") + (piece.color === void 0 ? "" : COLOR_TO_TERMINAL_ESCAPE[piece.color]) + piece.text + RESET_COLOR;
    case "Text":
      return piece.text;
  }
}
function fancyToPlainErrorLocation(location) {
  switch (location.tag) {
    case "ElmJsonPath":
      return { tag: "FileOnly", file: location.theElmJsonPath };
    case "ElmWatchJsonPath":
      return { tag: "FileOnly", file: location.theElmWatchJsonPath };
    case "ElmWatchStuffJsonPath":
      return { tag: "FileOnly", file: location.theElmWatchStuffJsonPath };
    case "OutputPath":
      return { tag: "Target", targetName: location.targetName };
    case "ElmWatchNodeScriptPath":
      return {
        tag: "FileOnly",
        file: markAsAbsolutePath(
          url.fileURLToPath(location.theElmWatchNodeScriptPath)
        )
      };
    case "FileWithLineAndColumn":
      return location;
    case "NoLocation":
      return void 0;
  }
}
function renderErrorLocation(location) {
  switch (location.tag) {
    case "FileOnly":
      return text(location.file);
    case "FileWithLineAndColumn":
      return text(`${location.file}:${location.line}:${location.column}`);
    case "Target":
      return dim2(`Target: ${location.targetName}`);
  }
}
function elmMakeJsonParseError(outputPath, error, errorFilePath, command) {
  return fancyError("TROUBLE WITH JSON REPORT", outputPath)`
I ran the following commands:

${printCommand(command)}

I seem to have gotten some JSON back as expected,
but I ran into an error when decoding it:

${printJsonError(error)}

${printErrorFilePath(errorFilePath)}
`;
}
function printCommand(command) {
  const stdin = command.stdin === void 0 ? "" : `${commandToPresentationName([
    "printf",
    truncate(command.stdin.toString("utf8"))
  ])} | `;
  return text(`
${commandToPresentationName(["cd", command.options.cwd])}
${stdin}${commandToPresentationName([command.command, ...command.args])}
`);
}
function commandToPresentationName(command) {
  return command.map(
    (part) => part === "" ? "''" : part.split(/(')/).map(
      (subPart) => subPart === "" ? "" : subPart === "'" ? "\\'" : /^[\w.,:/=@%+-]+$/.test(subPart) ? subPart : `'${subPart}'`
    ).join("")
  ).join(" ");
}
function printErrorFilePath(errorFilePath) {
  switch (errorFilePath.tag) {
    case "AbsolutePath":
      return template`
I wrote that to this file so you can inspect it:

${text(errorFilePath.theAbsolutePath)}
      `;
    case "WritingErrorFileFailed":
      return template`
I tried to write that to this file:

${text(errorFilePath.attemptedPath)}

${bold2("But that failed too:")}

${text(errorFilePath.error.message)}
      `;
    case "ErrorFileBadContent":
      return template`
I wrote this error to a file so you can inspect and possibly report it more easily.

This is the data that caused the error:

${text(errorFilePath.content)}
      `;
  }
}
function truncate(string7) {
  const roughLimit = 20;
  const half = Math.floor(roughLimit / 2);
  return string7.length <= roughLimit ? string7 : `${string7.slice(0, half)}...${string7.slice(-half)}`;
}
function printJsonError(error) {
  return text(Codec8.format(error));
}
function tryWriteErrorFile({
  cwd,
  name,
  content,
  hash
}) {
  const jsonPath = absolutePathFromString(
    cwd,
    `elm-watch-${name}-${sha256(hash)}.txt`
  );
  try {
    fs2.writeFileSync(jsonPath, content);
    return { tag: "AbsolutePath", theAbsolutePath: jsonPath };
  } catch (unknownError) {
    const error = toError(unknownError);
    return {
      tag: "WritingErrorFileFailed",
      error,
      attemptedPath: jsonPath
    };
  }
}
function sha256(string7) {
  return crypto.createHash("sha256").update(string7).digest("hex");
}

// src/Spawn.ts
import * as childProcess from "child_process";
function spawn2(command) {
  let killed = false;
  let kill = () => {
    killed = true;
  };
  const promise = (actualSpawn) => new Promise((resolve2) => {
    if (killed) {
      resolve2({ tag: "Killed", command });
      return;
    }
    const child = actualSpawn(command.command, command.args, {
      ...command.options,
      cwd: command.options.cwd
    });
    const stdout = [];
    const stderr = [];
    child.on("error", (error) => {
      resolve2(
        /* v8 ignore start */
        error.code === "ENOENT" ? { tag: "CommandNotFoundError", command } : { tag: "OtherSpawnError", error, command }
        /* v8 ignore stop */
      );
    });
    let stdinWriteError = void 0;
    child.stdin.on("error", (error) => {
      if (error.code === "EPIPE" || error.code === "EOF") {
        const result = {
          tag: "StdinWriteError",
          error,
          command
        };
        stdinWriteError = {
          result,
          timeoutId: setTimeout(() => {
            resolve2(result);
          }, 500)
        };
      } else {
        resolve2({ tag: "OtherSpawnError", error, command });
      }
    });
    child.stdout.on("error", (error) => {
      resolve2({ tag: "OtherSpawnError", error, command });
    });
    child.stderr.on("error", (error) => {
      resolve2({ tag: "OtherSpawnError", error, command });
    });
    child.stdout.on("data", (chunk) => {
      stdout.push(chunk);
    });
    child.stderr.on("data", (chunk) => {
      stderr.push(chunk);
    });
    child.on("exit", (exitCode, signal) => {
      if (killed) {
      } else {
        if (exitCode === 0 && stdinWriteError !== void 0) {
          clearTimeout(stdinWriteError.timeoutId);
          resolve2(stdinWriteError.result);
        } else {
          resolve2({
            tag: "Exit",
            exitReason: exitReason(exitCode, signal),
            stdout: Buffer.concat(stdout),
            stderr: Buffer.concat(stderr),
            command
          });
        }
      }
    });
    kill = () => {
      if (!killed) {
        child.kill();
        resolve2({ tag: "Killed", command });
        killed = true;
      }
    };
    if (command.stdin !== void 0) {
      child.stdin.end(command.stdin);
    }
  });
  return {
    promise: IS_WINDOWS ? import("cross-spawn").then((crossSpawn) => promise(crossSpawn.spawn)) : promise(childProcess.spawn),
    kill: () => {
      kill();
    }
  };
}
function exitReason(exitCode, signal) {
  return exitCode !== null ? { tag: "ExitCode", exitCode } : signal !== null ? { tag: "Signal", signal } : { tag: "Unknown" };
}

// src/SpawnElm.ts
function make({
  elmJsonPath,
  compilationMode,
  inputs,
  outputPath,
  env,
  getNow
}) {
  const cwd = absoluteDirname(elmJsonPath);
  const command = {
    command: "elm",
    args: [
      "make",
      "--report=json",
      ...maybeToArray(compilationModeToArg(compilationMode)),
      `--output=${outputPathToAbsoluteString(cwd, outputPath)}`,
      // Use relative paths because:
      // - Windows has a maximum command length. See: https://github.com/lydell/elm-watch/issues/86
      // - It looks nicer in error messages (the printed commands are much shorter).
      ...inputs.map((inputPath) => path2.relative(cwd, inputPath.theInputPath))
    ],
    options: {
      cwd,
      env
    }
  };
  const { promise, kill } = spawn2(command);
  const handleSpawnResult = (spawnResult) => {
    switch (spawnResult.tag) {
      /* v8 ignore start */
      case "CommandNotFoundError":
        return { tag: "ElmNotFoundError", command };
      /* v8 ignore stop */
      /* v8 ignore start */
      case "OtherSpawnError":
      case "StdinWriteError":
        return {
          tag: "OtherSpawnError",
          error: spawnResult.error,
          command: spawnResult.command
        };
      /* v8 ignore stop */
      case "Killed":
        return { tag: "Killed" };
      case "Exit": {
        const { exitReason: exitReason2 } = spawnResult;
        const stdout = spawnResult.stdout.toString("utf8");
        const stderr = spawnResult.stderr.toString("utf8");
        const unexpectedElmMakeOutput = {
          tag: "UnexpectedElmMakeOutput",
          exitReason: exitReason2,
          stdout,
          stderr,
          command
        };
        return exitReason2.tag === "ExitCode" && exitReason2.exitCode === 0 && stdout === "" && stderr === "" ? { tag: "Success" } : exitReason2.tag === "ExitCode" && exitReason2.exitCode === 1 && stdout === "" ? parsePotentialElmMakeJson(command, stderr) ?? unexpectedElmMakeOutput : unexpectedElmMakeOutput;
      }
    }
  };
  const startTime = getNow().getTime();
  return {
    promise: promise.then(handleSpawnResult),
    kill: ({ force }) => {
      if (force) {
        kill();
      } else {
        delayKill(promise, startTime, getNow, env, kill);
      }
    }
  };
}
function delayKill(promise, startTime, getNow, env, kill) {
  const timeout = silentlyReadIntEnvValue(
    env[__ELM_WATCH_ELM_TIMEOUT_MS],
    1e4
  );
  const elapsed = getNow().getTime() - startTime;
  const timeoutId = setTimeout(kill, Math.max(0, timeout - elapsed));
  void promise.finally(() => {
    clearTimeout(timeoutId);
  });
}
function compilationModeToArg(compilationMode) {
  switch (compilationMode) {
    case "standard":
      return void 0;
    case "debug":
      return "--debug";
    case "optimize":
      return "--optimize";
  }
}
function outputPathToAbsoluteString(cwd, outputPath) {
  switch (outputPath.tag) {
    case "OutputPath":
      return path2.relative(
        cwd,
        outputPath.writeToTemporaryDir ? outputPath.temporaryOutputPath : outputPath.theOutputPath
      );
    case "NullOutputPath":
      return "/dev/null";
  }
}
function maybeToArray(arg) {
  return arg === void 0 ? [] : [arg];
}
function parsePotentialElmMakeJson(command, stderr) {
  if (!stderr.endsWith("}")) {
    const errorIndex = stderr.lastIndexOf("elm: ");
    if (errorIndex !== -1) {
      return {
        tag: "ElmMakeCrashError",
        beforeError: stderr.startsWith("{") ? { tag: "Json", length: errorIndex } : { tag: "Text", text: stderr.slice(0, errorIndex) },
        error: stderr.slice(errorIndex),
        command
      };
    }
  }
  const match = elmStuffErrorMessagePrefixRegex.exec(stderr);
  const elmStuffError = match?.[0];
  const potentialJson = elmStuffError === void 0 ? stderr : stderr.slice(elmStuffError.length);
  return potentialJson.startsWith("{") ? parseActualElmMakeJson(command, potentialJson, elmStuffError?.trim()) : void 0;
}
function parseActualElmMakeJson(command, jsonString, extraError) {
  const parsed = Codec9.JSON.parse(
    ElmMakeError,
    jsonString.replace(/\t/g, "\\t")
  );
  switch (parsed.tag) {
    case "DecoderError":
      return {
        tag: "ElmMakeJsonParseError",
        error: parsed.error,
        errorFilePath: tryWriteErrorFile({
          cwd: command.options.cwd,
          name: "ElmMakeJsonParseError",
          content: toPlainString(
            elmMakeJsonParseError(
              { tag: "NoLocation" },
              parsed.error,
              { tag: "ErrorFileBadContent", content: jsonString },
              command
            )
          ),
          hash: jsonString
        }),
        command
      };
    case "Valid":
      return {
        tag: "ElmMakeError",
        error: parsed.value,
        extraError
      };
  }
}
var elmStuffErrorMessagePrefixRegex = /^\+-+\r?\n(?:\|.*\r?\n)+\+-+\r?\n\r?\n/;

// src/ElmJson.ts
import * as Codec10 from "tiny-decoders";
var ElmJson = Codec10.taggedUnion("type", [
  {
    type: Codec10.tag("application"),
    "source-directories": NonEmptyArray(Codec10.string)
  },
  {
    type: Codec10.tag("package")
  }
]);
function readSourceDirectories(elmJsonPath) {
  const parsed = readJsonFile(elmJsonPath, ElmJson);
  switch (parsed.tag) {
    case "DecoderError":
      return {
        tag: "ElmJsonDecodeError",
        elmJsonPath,
        error: parsed.error
      };
    case "ReadError":
      return {
        tag: "ElmJsonReadError",
        elmJsonPath,
        error: parsed.error
      };
    case "Valid":
      return {
        tag: "Parsed",
        sourceDirectories: getSourceDirectories(elmJsonPath, parsed.value)
      };
  }
}
function getSourceDirectories(elmJsonPath, elmJson2) {
  const base = absoluteDirname(elmJsonPath);
  switch (elmJson2.type) {
    case "application":
      return mapNonEmptyArray(
        elmJson2["source-directories"],
        (dir) => markAsSourceDirectory(absolutePathFromString(base, dir))
      );
    case "package":
      return [markAsSourceDirectory(absolutePathFromString(base, "src"))];
  }
}

// src/ImportWalker.ts
import * as fs3 from "fs";
import * as path3 from "path";

// src/Parser.ts
var MODULE_NAME = /^\p{Lu}[_\d\p{L}]*(?:\.\p{Lu}[_\d\p{L}]*)*$/u;
var CR = 13;
var HYPHEN = 45;
var LEFT_BRACE = 123;
var LF = 10;
var RIGHT_BRACE = 125;
var SPACE = 32;
function isImport(chars) {
  return chars.length === 6 && chars[0] === 105 && chars[1] === 109 && chars[2] === 112 && chars[3] === 111 && chars[4] === 114 && chars[5] === 116;
}
var initialReadState = () => ({
  tokenizerState: { tag: "Initial", chars: [], multilineCommentLevel: 0 },
  parserState: { tag: "StartOfFile" },
  importedModules: []
});
function readChar(char, readState) {
  const maybeToken = tokenize(char, readState.tokenizerState);
  if (maybeToken === void 0) {
    return;
  }
  const maybeModuleName = parse2(maybeToken, readState.parserState);
  if (maybeModuleName !== void 0) {
    readState.importedModules.push(maybeModuleName);
  }
}
function isNonImport(readState) {
  return readState.parserState.tag === "NonImport";
}
function finalize(readState) {
  if (readState.tokenizerState.tag !== "Initial") {
    return readState.importedModules;
  }
  const maybeLastToken = flush(readState.tokenizerState.chars);
  if (maybeLastToken === void 0) {
    return readState.importedModules;
  }
  const maybeLastModuleName = parse2(maybeLastToken, readState.parserState);
  if (maybeLastModuleName === void 0) {
    return readState.importedModules;
  }
  readState.importedModules.push(maybeLastModuleName);
  return readState.importedModules;
}
function tokenize(char, tokenizerState) {
  switch (tokenizerState.tag) {
    case "Initial":
      switch (char) {
        case SPACE: {
          const maybeToken = flush(tokenizerState.chars);
          tokenizerState.chars = [];
          return maybeToken;
        }
        case CR:
        case LF:
          tokenizerState.tag = "MaybeNewChunk";
          return flush(tokenizerState.chars);
        case LEFT_BRACE:
          tokenizerState.tag = "MaybeMultilineComment{";
          return flush(tokenizerState.chars);
        case HYPHEN:
          tokenizerState.tag = "MaybeSinglelineComment-";
          return flush(tokenizerState.chars);
        default:
          tokenizerState.chars.push(char);
          return void 0;
      }
    case "MaybeNewChunk":
      switch (char) {
        case SPACE:
          tokenizerState.tag = "Initial";
          tokenizerState.chars = [];
          return void 0;
        case CR:
        case LF:
          return void 0;
        case LEFT_BRACE:
          tokenizerState.tag = "MaybeMultilineComment{";
          return void 0;
        case HYPHEN:
          tokenizerState.tag = "MaybeSinglelineComment-";
          return void 0;
        default:
          tokenizerState.tag = "Initial";
          tokenizerState.chars = [char];
          return { tag: "NewChunk" };
      }
    case "MaybeMultilineComment{":
      switch (char) {
        case HYPHEN:
          tokenizerState.tag = "MultilineComment";
          tokenizerState.multilineCommentLevel = 1;
          return void 0;
        default:
          tokenizerState.tag = "Initial";
          tokenizerState.chars = [LEFT_BRACE];
          return tokenize(char, tokenizerState);
      }
    case "MultilineComment":
      switch (char) {
        case LEFT_BRACE:
          tokenizerState.tag = "MultilineComment{";
          return void 0;
        case HYPHEN:
          tokenizerState.tag = "MultilineComment-";
          return void 0;
        default:
          return void 0;
      }
    case "MultilineComment{":
      switch (char) {
        case HYPHEN:
          tokenizerState.tag = "MultilineComment";
          tokenizerState.multilineCommentLevel++;
          return void 0;
        case LEFT_BRACE:
          return void 0;
        default:
          tokenizerState.tag = "MultilineComment";
          return void 0;
      }
    case "MultilineComment-":
      switch (char) {
        case RIGHT_BRACE:
          if (tokenizerState.multilineCommentLevel <= 1) {
            tokenizerState.tag = "Initial";
            tokenizerState.chars = [];
          } else {
            tokenizerState.tag = "MultilineComment";
            tokenizerState.multilineCommentLevel--;
          }
          return void 0;
        case LEFT_BRACE:
          tokenizerState.tag = "MultilineComment{";
          return void 0;
        case HYPHEN:
          return void 0;
        default:
          tokenizerState.tag = "MultilineComment";
          return void 0;
      }
    case "MaybeSinglelineComment-":
      switch (char) {
        case HYPHEN:
          tokenizerState.tag = "SinglelineComment";
          return void 0;
        default:
          tokenizerState.tag = "Initial";
          tokenizerState.chars = [HYPHEN];
          return tokenize(char, tokenizerState);
      }
    case "SinglelineComment":
      switch (char) {
        case CR:
        case LF:
          tokenizerState.tag = "MaybeNewChunk";
          return void 0;
        default:
          return void 0;
      }
  }
}
function flush(chars) {
  return isNonEmptyArray(chars) ? { tag: "Word", chars } : void 0;
}
function parse2(token, parserState) {
  switch (parserState.tag) {
    case "StartOfFile":
      switch (token.tag) {
        case "NewChunk":
          return void 0;
        case "Word":
          if (isImport(token.chars)) {
            parserState.tag = "Import";
            return void 0;
          } else {
            parserState.tag = "Ignore";
            return void 0;
          }
      }
    case "NewChunk":
      switch (token.tag) {
        /* v8 ignore start */
        case "NewChunk":
          return void 0;
        /* v8 ignore stop */
        case "Word":
          if (isImport(token.chars)) {
            parserState.tag = "Import";
            return void 0;
          } else {
            parserState.tag = "NonImport";
            return void 0;
          }
      }
    case "Import":
      switch (token.tag) {
        case "NewChunk":
          parserState.tag = "NewChunk";
          return void 0;
        case "Word": {
          const string7 = Buffer.from(token.chars).toString();
          if (MODULE_NAME.test(string7)) {
            parserState.tag = "Ignore";
            return split(string7, ".");
          } else {
            parserState.tag = "Ignore";
            return void 0;
          }
        }
      }
    case "Ignore":
      switch (token.tag) {
        case "NewChunk":
          parserState.tag = "NewChunk";
          return void 0;
        case "Word":
          return void 0;
      }
    case "NonImport":
      return void 0;
  }
}

// src/ImportWalker.ts
function walkImports(sourceDirectories, inputRealPaths) {
  const allRelatedElmFilePaths = new Set(
    inputRealPaths.flatMap(
      (inputRealPath) => initialRelatedElmFilePaths(sourceDirectories, inputRealPath)
    )
  );
  const visitedModules = /* @__PURE__ */ new Set();
  try {
    for (const inputRealPath of inputRealPaths) {
      walkImportsHelper(
        mapNonEmptyArray(sourceDirectories, (sourceDirectory) => ({
          sourceDirectory,
          children: new Set(readdirSync2(sourceDirectory))
        })),
        inputRealPath,
        allRelatedElmFilePaths,
        visitedModules
      );
    }
  } catch (unknownError) {
    const error = toError(unknownError);
    return {
      tag: "ImportWalkerFileSystemError",
      error,
      relatedElmFilePathsUntilError: allRelatedElmFilePaths
    };
  }
  return { tag: "Success", allRelatedElmFilePaths };
}
function walkImportsHelper(sourceDirectories, elmFilePath, allRelatedElmFilePaths, visitedModules) {
  if (!fs3.existsSync(elmFilePath)) {
    return;
  }
  const importedModules = parse3(elmFilePath);
  for (const importedModule of importedModules) {
    const relativePath = `${importedModule.join(path3.sep)}.elm`;
    if (!visitedModules.has(relativePath)) {
      visitedModules.add(relativePath);
      for (const { sourceDirectory, children } of sourceDirectories) {
        const newElmFilePath = markAsAbsolutePath(
          sourceDirectory + path3.sep + relativePath
        );
        allRelatedElmFilePaths.add(newElmFilePath);
        const child = importedModule.length === 1 ? `${importedModule[0]}.elm` : importedModule[0];
        if (children.has(child)) {
          walkImportsHelper(
            sourceDirectories,
            newElmFilePath,
            allRelatedElmFilePaths,
            visitedModules
          );
        }
      }
    }
  }
}
function parse3(elmFilePath) {
  const readState = initialReadState();
  const handle = fs3.openSync(elmFilePath, "r");
  const buffer = Buffer.alloc(2048);
  let bytesRead = 0;
  outer: while ((bytesRead = fs3.readSync(handle, buffer)) > 0) {
    for (const char of buffer.subarray(0, bytesRead)) {
      readChar(char, readState);
      if (isNonImport(readState)) {
        break outer;
      }
    }
  }
  fs3.closeSync(handle);
  return finalize(readState);
}
function initialRelatedElmFilePaths(sourceDirectories, inputRealPath) {
  return [
    inputRealPath,
    ...sourceDirectories.flatMap((sourceDirectory) => {
      const prefix = sourceDirectory + path3.sep;
      return inputRealPath.startsWith(prefix) ? sourceDirectories.map(
        (sourceDirectory2) => markAsAbsolutePath(
          sourceDirectory2 + path3.sep + inputRealPath.slice(prefix.length)
        )
      ) : [];
    })
  ];
}
function readdirSync2(dir) {
  try {
    return fs3.readdirSync(dir);
  } catch {
    return [];
  }
}

// src/Inject.ts
import * as Codec11 from "tiny-decoders";
var REPLACEMENT_REGEX = /^(?:function (F|_Platform_initialize|_Platform_export|_Browser_application|_Scheduler_binding|_Scheduler_step)\(|var (_VirtualDom_init|\$elm\$browser\$Browser\$sandbox|_Platform_worker|_Browser_element|_Browser_document|_Debugger_element|_Debugger_document) =).*\r?\n?\{(?:.*\r?\n)*?\}\)?;?$/gm;
var PLACEHOLDER_REGEX = /%(\w+)%/g;
var REPLACEMENTS = {
  // ### _Platform_initialize (main : Program flags model msg)
  // New implementation.
  // Note: `isDebug` is needed when you have programs that do and don’t support
  // the debugger in the same output. `$elm$browser$Debugger$Main$wrapUpdate`
  // etc is going to be defined, but it should only be used in
  // `_Platform_initialize` when actually called from `_Debugger_element` or
  // `_Debugger_document`, not from `_Platform_worker`. (`Html` programs don’t
  // call `_Platform_initialize`.)
  _Platform_initialize: `
// added by elm-watch
var _elmWatchTargetName = "";

// This whole function was changed by elm-watch.
function _Platform_initialize(programType, isDebug, debugMetadata, flagDecoder, args, init, impl, stepperBuilder)
{
	if (args === "__elmWatchReturnData") {
		return { impl: impl, debugMetadata: debugMetadata, flagDecoder : flagDecoder, programType: programType, _Platform_effectManagers: _Platform_effectManagers, _Scheduler_enqueue: _Scheduler_enqueue };
	}

	var flags = _Json_wrap(args ? args['flags'] : undefined);
	var flagResult = A2(_Json_run, flagDecoder, flags);
	$elm$core$Result$isOk(flagResult) || _Debug_crash(2 /**/, _Json_errorToString(flagResult.a) /**/);
	var managers = {};
	var initUrl = programType === "Browser.application" ? _Browser_getUrl() : undefined;
	globalThis.__ELM_WATCH_INIT_URL = initUrl;
	var initPair = init(flagResult.a);
	delete globalThis.__ELM_WATCH_INIT_URL;
	var model = initPair.a;
	var stepper = stepperBuilder(sendToApp, model);
	var ports = _Platform_setupEffects(managers, sendToApp);
	var update;
	var subscriptions;

	function setUpdateAndSubscriptions() {
		update = impl.%update% || impl._impl.%update%;
		subscriptions = impl.%subscriptions% || impl._impl.%subscriptions%;
		if (isDebug) {
			update = $elm$browser$Debugger$Main$wrapUpdate(update);
			subscriptions = $elm$browser$Debugger$Main$wrapSubs(subscriptions);
		}
	}

	function sendToApp(msg, viewMetadata) {
		var pair = A2(update, msg, model);
		stepper(model = pair.a, viewMetadata);
		_Platform_enqueueEffects(managers, pair.b, subscriptions(model));
	}

	setUpdateAndSubscriptions();

	var skippedInitCmds = globalThis.__ELM_WATCH ? globalThis.__ELM_WATCH.SHOULD_SKIP_INIT_CMDS(_elmWatchTargetName) : false;

	_Platform_enqueueEffects(managers, skippedInitCmds ? _Platform_batch(_List_Nil) : initPair.b, subscriptions(model));

	function __elmWatchHotReload(newData) {
		_Platform_enqueueEffects(managers, _Platform_batch(_List_Nil), _Platform_batch(_List_Nil));
		_Scheduler_enqueue = newData._Scheduler_enqueue;

		var reloadReasons = [];

		for (var key in newData._Platform_effectManagers) {
			var manager = newData._Platform_effectManagers[key];
			if (!(key in _Platform_effectManagers)) {
				_Platform_effectManagers[key] = manager;
				managers[key] = _Platform_instantiateManager(manager, sendToApp);
				if (manager.a) {
					reloadReasons.push({ tag: "NewPortAdded", name: key });
					manager.a(key, sendToApp)
				}
			}
		}

		for (var key in newData.impl) {
			if (key === "_impl" && impl._impl) {
				for (var subKey in newData.impl[key]) {
					impl._impl[subKey] = newData.impl[key][subKey];
				}
			} else {
				impl[key] = newData.impl[key];
			}
		}

		var newFlagResult = A2(_Json_run, newData.flagDecoder, flags);
		if (!$elm$core$Result$isOk(newFlagResult)) {
			return reloadReasons.concat({ tag: "FlagsTypeChanged", jsonErrorMessage: _Json_errorToString(newFlagResult.a) });
		}
		if (!_Utils_eq_elmWatchInternal(debugMetadata, newData.debugMetadata)) {
			return reloadReasons.concat({ tag: "MessageTypeChangedInDebugMode" });
		}
		init = impl.%init% || impl._impl.%init%;
		if (isDebug) {
			init = A3($elm$browser$Debugger$Main$wrapInit, _Json_wrap(newData.debugMetadata), initPair.a.popout, init);
		}
		globalThis.__ELM_WATCH_INIT_URL = initUrl;
		var newInitPair = init(newFlagResult.a);
		delete globalThis.__ELM_WATCH_INIT_URL;
		if (!_Utils_eq_elmWatchInternal(initPair, newInitPair)) {
			return reloadReasons.concat({ tag: "InitReturnValueChanged" });
		}

		setUpdateAndSubscriptions();
		stepper(model, true /* isSync */);
		_Platform_enqueueEffects(managers, skippedInitCmds ? newInitPair.b : _Platform_batch(_List_Nil), subscriptions(model));
		skippedInitCmds = false;
		return reloadReasons;
	}

	function __elmWatchRunInitCmds() {
		if (skippedInitCmds) {
			_Platform_enqueueEffects(managers, initPair.b, subscriptions(model));
			skippedInitCmds = false;
		}
	}

	return Object.defineProperties(
		ports ? { ports: ports } : {},
		{
			__elmWatchHotReload: { value: __elmWatchHotReload },
			__elmWatchRunInitCmds: { value: __elmWatchRunInitCmds },
			__elmWatchProgramType: { value: programType },
		}
	);
}

// This whole function was added by elm-watch.
// Copy-paste of _Utils_eq but does not assume that x and y have the same type,
// and considers functions to always be equal.
function _Utils_eq_elmWatchInternal(x, y)
{
	for (
		var pair, stack = [], isEqual = _Utils_eqHelp_elmWatchInternal(x, y, 0, stack);
		isEqual && (pair = stack.pop());
		isEqual = _Utils_eqHelp_elmWatchInternal(pair.a, pair.b, 0, stack)
		)
	{}

	return isEqual;
}

// This whole function was added by elm-watch.
function _Utils_eqHelp_elmWatchInternal(x, y, depth, stack)
{
	if (x === y) {
		return true;
	}

	var xType = _Utils_typeof_elmWatchInternal(x);
	var yType = _Utils_typeof_elmWatchInternal(y);

	if (xType !== yType) {
		return false;
	}

	switch (xType) {
		case "primitive":
			return false;
		case "function":
			return true;
	}

	if (x.$ !== y.$) {
		return false;
	}

	if (x.$ === 'Set_elm_builtin') {
		x = $elm$core$Set$toList(x);
		y = $elm$core$Set$toList(y);
	} else if (x.$ === 'RBNode_elm_builtin' || x.$ === 'RBEmpty_elm_builtin' || x.$ < 0) {
		x = $elm$core$Dict$toList(x);
		y = $elm$core$Dict$toList(y);
	}

	if (Object.keys(x).length !== Object.keys(y).length) {
		return false;
	}

	if (depth > 100) {
		stack.push(_Utils_Tuple2(x, y));
		return true;
	}

	for (var key in x) {
		if (!_Utils_eqHelp_elmWatchInternal(x[key], y[key], depth + 1, stack)) {
			return false;
		}
	}
	return true;
}

// This whole function was added by elm-watch.
function _Utils_typeof_elmWatchInternal(x)
{
	var type = typeof x;
	return type === "function"
		? "function"
		: type !== "object" || type === null
		? "primitive"
		: "objectOrArray";
}
				`.trim(),
  // Make sure these are always defined for easier code in `_Platform_initialize`.
  // We don’t actually do anything with the `F` function – it’s just a way to get
  // these definitions near the top of the file.
  F: `
var _Platform_effectManagers = {}, _Scheduler_enqueue; // added by elm-watch

function F(arity, fun, wrapper) {
  wrapper.a = arity;
  wrapper.f = fun;
  return wrapper;
}
  `.trim(),
  // ### _VirtualDom_init (main : Html msg)
  // New implementation.
  _VirtualDom_init: `
// This whole function was changed by elm-watch.
var _VirtualDom_init = F4(function(virtualNode, flagDecoder, debugMetadata, args)
{
	var programType = "Html";

	if (args === "__elmWatchReturnData") {
		return { virtualNode: virtualNode, programType: programType, _Platform_effectManagers: _Platform_effectManagers, _Scheduler_enqueue: _Scheduler_enqueue };
	}

	/**_UNUSED/ // always UNUSED with elm-watch
	var node = args['node'];
	//*/
	/**/
	var node = args && args['node'] ? args['node'] : _Debug_crash(0);
	//*/

	var nextNode = _VirtualDom_render(virtualNode, function() {});
	node.parentNode.replaceChild(nextNode, node);
	node = nextNode;
	var sendToApp = function() {};

	function __elmWatchHotReload(newData) {
		var patches = _VirtualDom_diff(virtualNode, newData.virtualNode);
		node = _VirtualDom_applyPatches(node, virtualNode, patches, sendToApp);
		virtualNode = newData.virtualNode;
		return [];
	}

	return Object.defineProperties(
		{},
		{
			__elmWatchHotReload: { value: __elmWatchHotReload },
			__elmWatchRunInitCmds: { value: Function.prototype },
			__elmWatchProgramType: { value: programType },
		}
	);
});
				`.trim(),
  // ### _Platform_export
  // Allow registering the exports, and skipping the usual merging of `window.Elm` (for hot reloading).
  _Platform_export: `
function _Platform_export(exports)
{
	// added by elm-watch
	if (globalThis.__ELM_WATCH) {
		if (globalThis.__ELM_WATCH.IS_REGISTERING) {
			globalThis.__ELM_WATCH.REGISTER(_elmWatchTargetName, exports);
		} else {
			globalThis.__ELM_WATCH.HOT_RELOAD(_elmWatchTargetName, exports);
			return;
		}
	}

	scope['Elm']
		? _Platform_mergeExportsDebug('Elm', scope['Elm'], exports) // Always calls the debug version with elm-watch (the only difference is an error message).
		: scope['Elm'] = exports;
}
				`.trim(),
  // ### _Browser_application
  // Don’t pluck things out of `impl`. Pass `impl` to `_Browser_document`. Init
  // with URL given from `_Platform_initialize` (via `globalThis.__ELM_WATCH_INIT_URL`).
  _Browser_application: `
// This function was slightly modified by elm-watch.
function _Browser_application(impl)
{
	// var onUrlChange = impl.onUrlChange; // commented out by elm-watch
	// var onUrlRequest = impl.onUrlRequest; // commented out by elm-watch
	// var key = function() { key.a(onUrlChange(_Browser_getUrl())); }; // commented out by elm-watch
	var key = function() { key.a(impl.%onUrlChange%(_Browser_getUrl())); }; // added by elm-watch

	return _Browser_document({
		%setup%: function(sendToApp)
		{
			key.a = sendToApp;
			_Browser_window.addEventListener('popstate', key);
			_Browser_window.navigator.userAgent.indexOf('Trident') < 0 || _Browser_window.addEventListener('hashchange', key);

			return F2(function(domNode, event)
			{
				if (!event.ctrlKey && !event.metaKey && !event.shiftKey && event.button < 1 && !domNode.target && !domNode.hasAttribute('download'))
				{
					event.preventDefault();
					var href = domNode.href;
					var curr = _Browser_getUrl();
					var next = $elm$url$Url$fromString(href).a;
					sendToApp(impl.%onUrlRequest%(
						(next
							&& curr.%protocol% === next.%protocol%
							&& curr.%host% === next.%host%
							&& curr.%port_%.a === next.%port_%.a
						)
							? $elm$browser$Browser$Internal(next)
							: $elm$browser$Browser$External(href)
					));
				}
			});
		},
		%init%: function(flags)
		{
			// return A3(impl.init, flags, _Browser_getUrl(), key); // commented out by elm-watch
			return A3(impl.%init%, flags, globalThis.__ELM_WATCH_INIT_URL, key); // added by elm-watch
		},
		// view: impl.view, // commented out by elm-watch
		// update: impl.update, // commented out by elm-watch
		// subscriptions: impl.subscriptions // commented out by elm-watch
		%view%: function(model) { return impl.%view%(model); }, // added by elm-watch
		_impl: impl // added by elm-watch
	});
}
  `.trim(),
  // ### $elm$browser$Browser$sandbox
  // Don’t pluck `view` from `impl`. Pass `impl` to `_Browser_element`.
  $elm$browser$Browser$sandbox: `
// This function was slightly modified by elm-watch.
var $elm$browser$Browser$sandbox = function (impl) {
	return _Browser_element(
		{
			%init%: function (_v0) {
				return _Utils_Tuple2(impl.%init%, $elm$core$Platform$Cmd$none);
			},
			%subscriptions%: function (_v1) {
				return $elm$core$Platform$Sub$none;
			},
			%update%: F2(
				function (msg, model) {
					return _Utils_Tuple2(
						A2(impl.%update%, msg, model),
						$elm$core$Platform$Cmd$none);
				}),
			// view: impl.view // commented out by elm-watch
			%view%: function(model) { return impl.%view%(model); }, // added by elm-watch
			_impl: impl // added by elm-watch
		});
};
  `.trim(),
  // ### _Platform_worker, _Browser_element, _Browser_document, _Debugger_element, _Debugger_document
  // Update call to `_Platform_initialize` to match our implementation.
  // `_Browser_application` calls `_Browser_document`/`_Debugger_document`.
  // `$elm$browser$Browser$sandbox` calls `_Browser_element`/`_Debugger_element`.
  // In those cases we need `impl._impl`.
  // Don’t pluck `view` from `impl`.
  // Also pass the type of program, `isDebug` and the `debugMetadata` to `_Platform_initialize`.
  _Platform_worker: `
// This function was slightly modified by elm-watch.
var _Platform_worker = F4(function(impl, flagDecoder, debugMetadata, args)
{
	return _Platform_initialize(
		"Platform.worker", // added by elm-watch
		false, // isDebug, added by elm-watch
		debugMetadata, // added by elm-watch
		flagDecoder,
		args,
		impl.%init%,
		// impl.update, // commented out by elm-watch
		// impl.subscriptions, // commented out by elm-watch
		impl, // added by elm-watch
		function() { return function() {} }
	);
});
  `.trim(),
  _Browser_element: `
// This function was slightly modified by elm-watch.
var _Browser_element = _Debugger_element || F4(function(impl, flagDecoder, debugMetadata, args)
{
	return _Platform_initialize(
		impl._impl ? "Browser.sandbox" : "Browser.element", // added by elm-watch
		false, // isDebug, added by elm-watch
		debugMetadata, // added by elm-watch
		flagDecoder,
		args,
		impl.%init%,
		// impl.update, // commented out by elm-watch
		// impl.subscriptions, // commented out by elm-watch
		impl, // added by elm-watch
		function(sendToApp, initialModel) {
			// var view = impl.view; // commented out by elm-watch
			/**_UNUSED/ // always UNUSED with elm-watch
			var domNode = args['node'];
			//*/
			/**/
			var domNode = args && args['node'] ? args['node'] : _Debug_crash(0);
			//*/
			var currNode = _VirtualDom_virtualize(domNode);

			return _Browser_makeAnimator(initialModel, function(model)
			{
				// var nextNode = view(model); // commented out by elm-watch
				var nextNode = impl.%view%(model); // added by elm-watch
				var patches = _VirtualDom_diff(currNode, nextNode);
				domNode = _VirtualDom_applyPatches(domNode, currNode, patches, sendToApp);
				currNode = nextNode;
			});
		}
	);
});
  `.trim(),
  _Browser_document: `
// This function was slightly modified by elm-watch.
var _Browser_document = _Debugger_document || F4(function(impl, flagDecoder, debugMetadata, args)
{
	return _Platform_initialize(
		impl._impl ? "Browser.application" : "Browser.document", // added by elm-watch
		false, // isDebug, added by elm-watch
		debugMetadata, // added by elm-watch
		flagDecoder,
		args,
		impl.%init%,
		// impl.update, // commented out by elm-watch
		// impl.subscriptions, // commented out by elm-watch
		impl, // added by elm-watch
		function(sendToApp, initialModel) {
			var divertHrefToApp = impl.%setup% && impl.%setup%(sendToApp)
			// var view = impl.view; // commented out by elm-watch
			var title = _VirtualDom_doc.title;
			var bodyNode = _VirtualDom_doc.body;
			_VirtualDom_divertHrefToApp = divertHrefToApp; // added by elm-watch
			var currNode = _VirtualDom_virtualize(bodyNode);
			_VirtualDom_divertHrefToApp = 0; // added by elm-watch
			return _Browser_makeAnimator(initialModel, function(model)
			{
				_VirtualDom_divertHrefToApp = divertHrefToApp;
				// var doc = view(model); // commented out by elm-watch
				var doc = impl.%view%(model); // added by elm-watch
				var nextNode = _VirtualDom_node('body')(_List_Nil)(doc.%body%);
				var patches = _VirtualDom_diff(currNode, nextNode);
				bodyNode = _VirtualDom_applyPatches(bodyNode, currNode, patches, sendToApp);
				currNode = nextNode;
				_VirtualDom_divertHrefToApp = 0;
				(title !== doc.%title%) && (_VirtualDom_doc.title = title = doc.%title%);
			});
		}
	);
});
  `.trim(),
  // Note: Debugger code does not need to worry about optimize mode shortened record fields.
  _Debugger_element: `
// This function was slightly modified by elm-watch.
var _Debugger_element = F4(function(impl, flagDecoder, debugMetadata, args)
{
	return _Platform_initialize(
		impl._impl ? "Browser.sandbox" : "Browser.element", // added by elm-watch
		true, // isDebug, added by elm-watch
		debugMetadata, // added by elm-watch
		flagDecoder,
		args,
		A3($elm$browser$Debugger$Main$wrapInit, _Json_wrap(debugMetadata), _Debugger_popout(), impl.init),
		// $elm$browser$Debugger$Main$wrapUpdate(impl.update), // commented out by elm-watch
		// $elm$browser$Debugger$Main$wrapSubs(impl.subscriptions), // commented out by elm-watch
		impl, // added by elm-watch
		function(sendToApp, initialModel)
		{
			// var view = impl.view; // commented out by elm-watch
			var title = _VirtualDom_doc.title;
			var domNode = args && args['node'] ? args['node'] : _Debug_crash(0);
			var currNode = _VirtualDom_virtualize(domNode);
			var currBlocker = $elm$browser$Debugger$Main$toBlockerType(initialModel);
			var currPopout;

			var cornerNode = _VirtualDom_doc.createElement('div');
			domNode.parentNode.insertBefore(cornerNode, domNode.nextSibling);
			var cornerCurr = _VirtualDom_virtualize(cornerNode);

			initialModel.popout.a = sendToApp;

			return _Browser_makeAnimator(initialModel, function(model)
			{
				// var nextNode = A2(_VirtualDom_map, $elm$browser$Debugger$Main$UserMsg, view($elm$browser$Debugger$Main$getUserModel(model))); // commented out by elm-watch
				var nextNode = A2(_VirtualDom_map, $elm$browser$Debugger$Main$UserMsg, impl.view($elm$browser$Debugger$Main$getUserModel(model))); // added by elm-watch
				var patches = _VirtualDom_diff(currNode, nextNode);
				domNode = _VirtualDom_applyPatches(domNode, currNode, patches, sendToApp);
				currNode = nextNode;

				// update blocker

				var nextBlocker = $elm$browser$Debugger$Main$toBlockerType(model);
				_Debugger_updateBlocker(currBlocker, nextBlocker);
				currBlocker = nextBlocker;

				// view corner

				var cornerNext = $elm$browser$Debugger$Main$cornerView(model);
				var cornerPatches = _VirtualDom_diff(cornerCurr, cornerNext);
				cornerNode = _VirtualDom_applyPatches(cornerNode, cornerCurr, cornerPatches, sendToApp);
				cornerCurr = cornerNext;

				if (!model.popout.b)
				{
					currPopout = undefined;
					return;
				}

				// view popout

				_VirtualDom_doc = model.popout.b; // SWITCH TO POPOUT DOC
				// currPopout || (currPopout = _VirtualDom_virtualize(model.popout.b)); // commented out by elm-watch
				currPopout || (currPopout = _VirtualDom_virtualize(model.popout.b.body)); // added by elm-watch
				var nextPopout = $elm$browser$Debugger$Main$popoutView(model);
				var popoutPatches = _VirtualDom_diff(currPopout, nextPopout);
				_VirtualDom_applyPatches(model.popout.b.body, currPopout, popoutPatches, sendToApp);
				currPopout = nextPopout;
				_VirtualDom_doc = document; // SWITCH BACK TO NORMAL DOC
			});
		}
	);
});
  `.trim(),
  _Debugger_document: `
// This function was slightly modified by elm-watch.
var _Debugger_document = F4(function(impl, flagDecoder, debugMetadata, args)
{
	return _Platform_initialize(
		impl._impl ? "Browser.application" : "Browser.document", // added by elm-watch
		true, // isDebug, added by elm-watch
		debugMetadata, // added by elm-watch
		flagDecoder,
		args,
		A3($elm$browser$Debugger$Main$wrapInit, _Json_wrap(debugMetadata), _Debugger_popout(), impl.init),
		// $elm$browser$Debugger$Main$wrapUpdate(impl.update), // commented out by elm-watch
		// $elm$browser$Debugger$Main$wrapSubs(impl.subscriptions), // commented out by elm-watch
		impl, // added by elm-watch
		function(sendToApp, initialModel)
		{
			var divertHrefToApp = impl.setup && impl.setup(function(x) { return sendToApp($elm$browser$Debugger$Main$UserMsg(x)); });
			// var view = impl.view; // commented out by elm-watch
			var title = _VirtualDom_doc.title;
			var bodyNode = _VirtualDom_doc.body;
			_VirtualDom_divertHrefToApp = divertHrefToApp; // added by elm-watch
			var currNode = _VirtualDom_virtualize(bodyNode);
			_VirtualDom_divertHrefToApp = 0; // added by elm-watch
			var currBlocker = $elm$browser$Debugger$Main$toBlockerType(initialModel);
			var currPopout;

			initialModel.popout.a = sendToApp;

			return _Browser_makeAnimator(initialModel, function(model)
			{
				_VirtualDom_divertHrefToApp = divertHrefToApp;
				// var doc = view($elm$browser$Debugger$Main$getUserModel(model)); // commented out by elm-watch
				var doc = impl.view($elm$browser$Debugger$Main$getUserModel(model)); // added by elm-watch
				var nextNode = _VirtualDom_node('body')(_List_Nil)(
					_Utils_ap(
						A2($elm$core$List$map, _VirtualDom_map($elm$browser$Debugger$Main$UserMsg), doc.body),
						_List_Cons($elm$browser$Debugger$Main$cornerView(model), _List_Nil)
					)
				);
				var patches = _VirtualDom_diff(currNode, nextNode);
				bodyNode = _VirtualDom_applyPatches(bodyNode, currNode, patches, sendToApp);
				currNode = nextNode;
				_VirtualDom_divertHrefToApp = 0;
				(title !== doc.title) && (_VirtualDom_doc.title = title = doc.title);

				// update blocker

				var nextBlocker = $elm$browser$Debugger$Main$toBlockerType(model);
				_Debugger_updateBlocker(currBlocker, nextBlocker);
				currBlocker = nextBlocker;

				// view popout

				if (!model.popout.b) { currPopout = undefined; return; }

				_VirtualDom_doc = model.popout.b; // SWITCH TO POPOUT DOC
				// currPopout || (currPopout = _VirtualDom_virtualize(model.popout.b)); // commented out by elm-watch
				currPopout || (currPopout = _VirtualDom_virtualize(model.popout.b.body)); // added by elm-watch
				var nextPopout = $elm$browser$Debugger$Main$popoutView(model);
				var popoutPatches = _VirtualDom_diff(currPopout, nextPopout);
				_VirtualDom_applyPatches(model.popout.b.body, currPopout, popoutPatches, sendToApp);
				currPopout = nextPopout;
				_VirtualDom_doc = document; // SWITCH BACK TO NORMAL DOC
			});
		}
	);
});
  `.trim(),
  // ### _Scheduler_binding, _Scheduler_step
  // This is needed because Elm mutates `Task`s in `_Scheduler_step`:
  //
  //     proc.__root.__kill = proc.__root.__callback(function(newRoot) {
  //
  // Some tasks are cancelable so `.__kill` is set to a function. Some are not,
  // and then `.__kill` seems to be set to `undefined`. But the initial value is
  // `null`! Later, there's just a truthiness check on `.__kill` so both `null` and
  // `undefined` works. However, this means that the “did `init` return the same
  // thing as last time?” check fails:
  //
  // - Either because of `null` vs `undefined`.
  // - Or because of `null` vs `function`.
  //
  // To solve this, make sure that `.__kill` (called `.c` in what we have to
  // work with below) is _always_ set to a function – a dummy one for
  // non-cancelable tasks (`Function.prototype` is a no-op function).
  // `_Utils_eq_elmWatchInternal` considers all functions to be equal.
  _Scheduler_binding: `
// This function was slightly modified by elm-watch.
function _Scheduler_binding(callback)
{
	return {
		$: 2,
		b: callback,
		// c: null // commented out by elm-watch
		c: Function.prototype // added by elm-watch
	};
}
  `.trim(),
  _Scheduler_step: `
function _Scheduler_step(proc)
{
	while (proc.f)
	{
		var rootTag = proc.f.$;
		if (rootTag === 0 || rootTag === 1)
		{
			while (proc.g && proc.g.$ !== rootTag)
			{
				proc.g = proc.g.i;
			}
			if (!proc.g)
			{
				return;
			}
			proc.f = proc.g.b(proc.f.a);
			proc.g = proc.g.i;
		}
		else if (rootTag === 2)
		{
			proc.f.c = proc.f.b(function(newRoot) {
				proc.f = newRoot;
				_Scheduler_enqueue(proc);
			// }); // commented out by elm-watch
			}) || Function.prototype; // added by elm-watch
			return;
		}
		else if (rootTag === 5)
		{
			if (proc.h.length === 0)
			{
				return;
			}
			proc.f = proc.f.b(proc.h.shift());
		}
		else // if (rootTag === 3 || rootTag === 4)
		{
			proc.g = {
				$: rootTag === 3 ? 0 : 1,
				b: proc.f.b,
				i: proc.g
			};
			proc.f = proc.f.d;
		}
	}
}
  `.trim()
};
var REPLACEMENTS_WITHOUT_PLACEHOLDERS = updateReplacements({}, REPLACEMENTS);
function inject(compilationMode, code, targetName) {
  const replacements = getReplacements(compilationMode, code);
  return code.replace(
    REPLACEMENT_REGEX,
    (match, name1, name = name1) => (
      /* v8 ignore start */
      replacements[name] ?? `${match} /* elm-watch ERROR: No replacement for function '${name}' was found! */`
    )
    /* v8 ignore stop */
  ).replace(
    /^var _elmWatchTargetName = "";$/m,
    /* v8 ignore next */
    `var _elmWatchTargetName = ${Codec11.JSON.stringify(Codec11.string, targetName ?? "")};`
  );
}
function getReplacements(compilationMode, code) {
  switch (compilationMode) {
    case "debug":
    case "standard":
      return REPLACEMENTS_WITHOUT_PLACEHOLDERS;
    case "optimize":
      return updateReplacements(getOptimizeModeRecordNames(code), REPLACEMENTS);
  }
}
function getOptimizeModeRecordNames(code) {
  const match1 = /^\s*impl\.([\w$]+),\s*impl\.([\w$]+),\s*impl\.([\w$]+),/m.exec(code);
  const match2 = /^\s*var divertHrefToApp = impl\.([\w$]+)/m.exec(code);
  const match3 = /^\s*var nextNode = _VirtualDom_node\('body'\)\(_List_Nil\)\(doc\.([\w$]+)\);/m.exec(
    code
  );
  const match4 = /^\s*\(title !== doc\.([\w$]+)\)/m.exec(code);
  const match5 = /^\s*&& curr\.([\w$]+) .*\s*&& curr\.([\w$]+) .*\s*&& curr\.([\w$]+)\..*/m.exec(
    code
  );
  const [
    ,
    init = "init_missing",
    update = "update_missing",
    subscriptions = "subscriptions_missing"
  ] = match1 ?? [];
  const [, setup = "setup_missing"] = match2 ?? [];
  const [, body = "body_missing"] = match3 ?? [];
  const [, title = "title_missing"] = match4 ?? [];
  const [
    ,
    protocol = "protocol_missing",
    host = "host_missing",
    port_ = "port__missing"
  ] = match5 ?? [];
  const extra = Object.fromEntries(
    Array.from(
      code.matchAll(/^\s*var ([\w$]+) = impl\.([\w$]+);/gm),
      /* v8 ignore next */
      ([, from = "from_missing", to = "to_missing"]) => [from, to]
    )
  );
  return {
    ...extra,
    init,
    update,
    subscriptions,
    setup,
    body,
    title,
    protocol,
    host,
    port_
  };
}
function updateReplacements(optimizeModeRecordNames, replacements) {
  return Object.fromEntries(
    Object.entries(replacements).map(([key, value]) => [
      key,
      updateString(optimizeModeRecordNames, value)
    ])
  );
}
function updateString(optimizeModeRecordNames, string7) {
  return string7.replace(
    PLACEHOLDER_REGEX,
    (_, name) => optimizeModeRecordNames[name] ?? name
  );
}

// src/elm-watch-lib.ts
function elmMake({
  elmJsonPath,
  compilationMode,
  inputs,
  outputPath,
  env
}) {
  return make({
    elmJsonPath: markAsElmJsonPath(markAsAbsolutePath(elmJsonPath)),
    compilationMode,
    inputs: mapNonEmptyArray(inputs, (rawInput) => ({
      tag: "InputPath",
      theInputPath: markAsAbsolutePath(rawInput)
    })),
    outputPath: outputPath === "/dev/null" ? { tag: "NullOutputPath" } : {
      tag: "OutputPath",
      theOutputPath: markAsAbsolutePath(outputPath),
      temporaryOutputPath: markAsAbsolutePath(outputPath),
      writeToTemporaryDir: false
    },
    env,
    getNow: () => /* @__PURE__ */ new Date()
  });
}
export {
  elmMake,
  inject,
  readSourceDirectories,
  walkImports
};
